File: longest-increasing-subsequence/flynn.cpp
/**
 * 풀이
 * - 아래와 같은 배열 memo를 이용하여 이중 반복문을 실행하는 풀이입니다
 * - memo[i]: nums[i]로 끝나는 subsequence 중에서 길이가 가장 긴 subsequence의 길이
 * 
 * Big O
 * - N: 배열 nums의 길이
 * - Time complexity: O(N^2)
 * - Space complexity: O(N)
 */

class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> memo;
        memo.push_back(1);

        int res = 1;
        for (int i = 1; i < nums.size(); ++i) {
            int tmp = 1;
            for (int j = 0; j < i; ++j) {
                if (nums[j] < nums[i]) tmp = max(tmp, memo[j] + 1);
            }
            memo.push_back(tmp);
            res = max(res, tmp);
        }

        return res;
    }
};

/**
 * 풀이
 * - wikipedia의 pseudo code를 참고하였습니다
 *   달레님 블로그에 실린 풀이를 통해서 훨씬 간단하게 문제에서 요구하는 바를 구할 수 있으므로, 문제의 풀이만을 원하신다면 달레님 블로그를 추천드리고
 *   좀 더 general하고 확장성 있는 알고리즘에 대해 궁금하신 분들께서는 wiki도 읽어보시는 걸 추천드립니다 (이해하는 데에는 시간이 좀 걸렸습니다)
 *   
 *   제가 읽고 이해한 결과 wiki 풀이와 달레님 풀이의 비교는 다음과 같습니다
 *   
 *   공통점: 문제에서 요구하는 바를 구할 수 있음 (LIS의 길이)
 *   차이점: wiki 풀이는 문제에서 요구하는 바를 구하는 것에 비해 overkill입니다 (이 문제에서는 굳이 필요 없는 부분이 꽤 있음)
 *         대신, wiki 풀이는 확장성이 좀 더 넓은 풀이입니다 (각 길이에 해당하는 increasing subsequence를 재구축할 수 있음)  
 * 
 *   관심 있으신 분들께서는 한 번 읽어보시는 것을 추천합니다 :)
 * - 참고: https://en.wikipedia.org/wiki/Longest_increasing_subsequence#Efficient_algorithms
 * 
 * - memo[l]: 현재 nums[i]를 탐색중이라고 할 때, l <= i인 l에 대하여
 *            길이가 l인 increasing subsequence들의 마지막 원소 중에서
 *            가장 최소값인 nums[k]의 인덱스 k
 *            nums를 순차적으로 탐색해 나감에 따라 계속 갱신되며 정렬된 상태를 유지하게 됩니다 (if x < y then nums[memo[x]] < nums[memo[y]])
 * 
 * - predec[i]: nums[i]를 마지막 원소로 하는 가장 긴 increasing subsequence에서 nums[i] 바로 앞에 오는 원소의 index
 * 
 * - nums를 순차적으로 탐색하며, 현재 탐색 중인 nums[i]를 마지막 원소로 삼는 가장 긴 Increasing subsequence를 찾습니다
 *   가장 긴 Increasing subsequence는 memo 배열에 대해 이분탐색을 실행하여 알아낼 수 있습니다
 * 
 * Big O
 * - N: 배열 nums의 길이
 * 
 * - Time complexity: O(NlogN)
 *   - nums의 각 원소마다 memo에 대해 이분탐색을 실행하므로 N이 증가함에 따라 실행 시간은 N * logN 형태로 증가합니다
 * - Space complexity: O(N)
 *   - memo 배열의 크기는 N이 증가함에 따라 선형적으로 증가합니다
 *  (- predec 배열의 크기 또한 N이 증가함에 따라 선형적으로 증가합니다)
 */

class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();

        vector<int> memo(n + 1, -1);

        // vector<int> predec(n, -1);
        // 각 길이에 맞는 increasing subsequence를 재구축할 때 쓰입니다

        int max_len = 0;
        for (int i = 0; i < nums.size(); ++i) {
            int lo = 1;
            int hi = max_len + 1;
            while (lo < hi) {
                int mid = lo + (hi - lo) / 2;
                if (nums[memo[mid]] < nums[i]) lo = mid + 1;
                else hi = mid;
            }
            // 위 이분탐색을 마치면 lo == hi인데
            // lo (혹은 hi)가 의미하는 바는 `nums[i]가 마지막 원소인 increasing subsequence 중에 길이가 가장 긴 녀석의 길이` 입니다
            int curr_len = lo;
            // 이해하기 쉽게끔 curr_len이라는 변수를 선언해줍니다

            // predec[i] = memo[curr_len - 1];
            // nums[i]가 마지막으로 오는 가장 긴 increasing subsequence에서 nums[i]의 바로 전 원소의 index를 기록해줍니다
            // 
            memo[curr_len] = i;

            if (curr_len > max_len) {
                // 만약 이전까지 찾았던 max_len보다 더 길이가 긴 increasing subsequence를 max_len
                max_len = curr_len;
            }
        }

        return max_len;

    // 길이 L짜리 increasing subsequence 중 하나를 재구축하려면 아래처럼 하면 됩니다
    // [P...P[memo[L]], ..., P[P[memo[L]]], P[memo[L]] ,memo[L]]

    //     vector<int> s(L, -1);
    //     int k = memo[L];
    //     for (int i = L - 1; i >= 0; --i) {
    //         s[i] = nums[k];
    //         k = predec[k];
    //     }
    }
};

File: find-minimum-in-rotated-sorted-array/flynn.go
/*
풀이 1
- 주어진 배열을 두 부분으로 나눌 수 있기 때문에 이진탐색을 이용하여 풀이할 수 있습니다
  주어진 배열이 A = [4,5,6,7,0,1,2] 라고 할 때, 이 배열은 두 부분으로 나눌 수 있습니다
    A[0:3] : rotate되어 앞으로 배열의 앞으로 옮겨진 부분
	A[4:6] : rotate되어 배열의 뒤로 밀려난 부분
  이걸 조금 다르게 표현하면 이렇게도 바라볼 수 있습니다
  f(A) = [T, T, T, T, F, F, F] (T/F: rotate되어 배열의 앞으로 옮겨진 부분인가?)

  이 때, 우리가 찾는 값 (the Minimum in the rotated sorted array)는 두 구간의 경계에 위치하고 있습니다
  즉, 첫번째 F의 위치를 찾는 문제로 바꿔 생각할 수 있단 뜻입니다

Big O
- N: 주어진 배열 nums의 길이

- Time complexity: O(logN)
- Space complexity: O(1)
*/

func findMin(nums []int) int {
	lo, hi := 0, len(nums)-1
	// rotate가 0번 실행된 경우, 이진탐색을 실행할 필요가 없고 이진탐색의 초기값 설정이 까다로워지기 때문에 처음에 따로 처리해줍니다
	// 앞서 hi의 초기값을 설정할 때, len(nums)가 아닌 len(nums) - 1로 설정할 수 있었던 이유이기도 합니다
	if nums[lo] < nums[hi] {
		return nums[lo]
	}

	// Go는 while문에 대응하는 표현을 for로 이렇게 표현합니다
	for lo < hi {
		mid := lo + (hi-lo)/2

		// 만일 mid가 앞으로 밀려난 부분에 속한다면...
		if nums[lo] <= nums[mid] && nums[mid] > nums[hi] {
			lo = mid + 1
		} else {
			hi = mid
		}
	}

	return nums[lo]
}

/*
풀이 2
- 풀이 1에서 덜어낼 수 있는 로직을 덜어낸 좀 더 깔끔한 이진탐색 풀이입니다

Big O
- 풀이 1과 동일
*/

func findMin(nums []int) int {
    lo, hi := 0, len(nums)

    for lo < hi {
        mid := lo+(hi-lo)/2

        if nums[mid] > nums[len(nums)-1] {
            lo = mid + 1
        } else {
            hi = mid
        }
    }

    return nums[lo]
}
